### PR TITLE
Namereg rebase

### PR DESCRIPTION
**Pull request by ebuchman**
A rebase of https://github.com/tendermint/tendermint/pull/77 on top of https://github.com/tendermint/tendermint/pull/82. I guess this can supplant both of them.

### CODE REVIEW

File: consensus/types/proposal.go [hunk 1]
```diff
@@ -38,9 +38,9 @@ func (p *Proposal) String() string {
 		p.BlockParts, p.POLParts, p.Signature)
 }
 
-func (p *Proposal) WriteSignBytes(w io.Writer, n *int64, err *error) {
-	// We hex encode the network name so we don't deal with escaping issues.
-	binary.WriteTo([]byte(Fmt(`{"network":"%X"`, config.GetString("network"))), w, n, err)
+func (p *Proposal) WriteSignBytes(chainID string, w io.Writer, n *int64, err *error) {
+	// We hex encode the chain_id name so we don't deal with escaping issues.
+	binary.WriteTo([]byte(Fmt(`{"chain_id":"%X"`, chainID)), w, n, err)
```
Comment (ebuchman) [inline consensus/types/proposal.go]: should be %s

File: rpc/core/txs.go [hunk 2]
```diff
@@ -91,27 +91,27 @@ func SignTx(tx types.Tx, privAccounts []*account.PrivAccount) (*ctypes.ResponseS
 		sendTx := tx.(*types.SendTx)
 		for i, input := range sendTx.Inputs {
 			input.PubKey = privAccounts[i].PubKey
-			input.Signature = privAccounts[i].Sign(sendTx)
+			input.Signature = privAccounts[i].Sign(config.GetString("chain_id"), sendTx)
```
Comment (ebuchman) [inline rpc/core/txs.go]: maybe this should be passed into the rpc call, and defaults to the config if empty.

File: rpc/core_client/client.go [hunk 3]
```diff
@@ -30,6 +28,7 @@ var reverseFuncMap = map[string]string{
 	"DumpStorage":    "dump_storage",
 	"BroadcastTx":    "broadcast_tx",
 	"ListAccounts":   "list_accounts",
+	"NameRegEntry":   "name_reg_entry",
```
Comment (ebuchman) [inline rpc/core_client/client.go]: needs an rpc-gen update

File: state/execution.go [hunk 4]
```diff
@@ -296,7 +296,7 @@ func ExecTx(blockCache *BlockCache, tx_ types.Tx, runCall bool, evc events.Firea
 
 	// TODO: do something with fees
 	fees := uint64(0)
-	_s := blockCache.State() // hack to access validators and event switch.
+	_s := blockCache.State() // hack to access validators and block height
```
Comment (ebuchman) [inline state/execution.go]: and chain id. is this really a hack still?

File: types/names.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+// base cost is "effective" number of bytes
+func BaseEntryCost(name, data string) uint64 {
+	return uint64(len(data) + 32)
+}
+
+type NameRegEntry struct {
+	Name    string `json:"name"`    // registered name for the entry
+	Owner   []byte `json:"owner"`   // address that created the entry
+	Data    string `json:"data"`    // binary encoded byte array
```
Comment (ebuchman) [inline types/names.go]: comment needs update

File: types/tx.go [hunk 6]
```diff
@@ -178,6 +183,54 @@ func (tx *CallTx) String() string {
...
+type NameTx struct {
+	Input *TxInput `json:"input"`
+	Name  string   `json:"name"`
+	Data  string   `json:"data"`
+	Fee   uint64   `json:"fee"`
+}
+
+func (tx *NameTx) WriteSignBytes(chainID string, w io.Writer, n *int64, err *error) {
+	// We hex encode the network name so we don't deal with escaping issues.
+	binary.WriteTo([]byte(Fmt(`{"network":"%X"`, chainID)), w, n, err)
```
Comment (ebuchman) [inline types/tx.go]: chain_id !

File: state/execution.go (hunk 4 above)
Comment (jaekwon) [inline state/execution.go]: It's like, the BlockCache is meant to act as a cache wrapper around the state, yet we mutate the validators directly.  We need to be careful, e.g. make sure we don't introduce a cache invalidation bug.  This should be documented for posterity.

File: types/tx.go [hunk 7]
```diff
@@ -178,6 +181,53 @@ func (tx *CallTx) String() string {
...
 //-----------------------------------------------------------------------------
 
+type NameTx struct {
+	Input *TxInput `json:"input"`
+	Name  string   `json:"name"`
+	Data  string   `json:"data"`
+	Fee   uint64   `json:"fee"`
+}
+
+func (tx *NameTx) WriteSignBytes(chainID string, w io.Writer, n *int64, err *error) {
```
Comment (jaekwon) [inline types/tx.go]: keys should be in alpha order... i'll make this change.

**Pull request merged**